cibyl/features/__init__.py

```diff
@@ -0,0 +1,156 @@
+
+    def __init__(self):
+        self.features_by_category = defaultdict(list)
+        self.all_features = {}
+        for location in self.features_locations:
+            for module_path in os.listdir(location):
+                module_match = MODULE_PATTERN.match(module_path)
+                if not module_match:
+                    continue
+                module_name = module_match.group(1)
+                # import the module directly from file, and add the _features
+                # suffix to the imported modules to help distinguish the
+                # classes defined there
+                # https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly
+                spec = spec_from_file_location(module_name+"_features",
+                                               f"{location}/{module_path}")
+                module = module_from_spec(spec)
+                spec.loader.exec_module(module)
+                features = getmembers(module, predicate=is_feature_class)
+                for feature_name, feature in features:
+                    self.all_features[feature_name.lower()] = feature
+                    self.features_by_category[module_name].append(feature_name)
+
+    def get_string_all_features(self):
+        """Get a string representation listing all available features to use in
+        an exception message."""
+        msg = ""
+        for category, features_names in self.features_by_category.items():
+            msg += f"\n{Colors.blue(category.title())}:"
+            for feature_name in features_names:
+                feature = self.all_features[feature_name.lower()]
+                docstring = getattr(feature, "__doc__", "")
+                msg += f"\n{Colors.red(' * ')}{Colors.blue(feature_name)}"
+                if docstring:
+                    msg += f"{Colors.red(' - '+feature.__doc__)}"
+        return msg
+
+    def get_feature(self, name_feature):
+        """Get the function associated with the given feature name
+
+        :param feature_name: Name of the feature
+        :type feature_name: str
+
+        :returns: class that implements the given feature
+        :rtype: class
+        :raises: InvalidArgument
+        """
+        try:
+            feature_class = self.all_features[name_feature.lower()]
+        except KeyError as err:
+            msg = f"No feature {name_feature}. Choose one of the following "
+            msg += "features:\n"
+            msg += self.get_string_all_features()
+            raise InvalidArgument(msg) from err
+        return feature_class()
+
+
+class FeatureTemplate(ABC):
+    """Skeleton for a generic feature."""
+    method_to_query = ""
+
+    @abstractmethod
+    def __init__(self, name):
+        self.name = name
+        self.args = {}
+
+    def query(self, system, **kwargs):
+        debug = kwargs.get("debug", False)
+        self.args.update(system.export_attributes_to_source())
+        self.args.update(kwargs)
```

Comment:
Have an abstract method where the inheritor has not provide their set of arguments instead. It will be much more obvious and I think that is what you meant with the abstract constructor. You can remove 'self.args' with that change.

Reply:
Yes, that looks cleaner, thanks